### PR TITLE
User can buy a subscription

### PR DESCRIPTION
**Pull request by yasmineezequiel**
[PT](https://www.pivotaltracker.com/story/show/169253720)
## PR Content
- Stripe as payment method
- Subscription controller 
- Adds user role and enables changing role 

## What we have learned 

- How to update change of model

### CODE REVIEW

**Review by miyeaier — APPROVED** [review]

File: app/controllers/api/v1/subscriptions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+      rescue => error
+        render_error(error.message)
+      end
+    else
+      render_error('No stripe token detected')
+    end
+  end
+
+  def new; end
```
Comment (oliverochman) [inline app/controllers/api/v1/subscriptions_controller.rb]: do you need to have the new action defined?

File: config/routes.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@
   namespace :api do
     namespace :v1 do
       resources :articles, only: [:index, :create, :update], constraints: { format: 'json' }
+      resources :subscriptions, only: [:new, :create, :index]
```
Comment (oliverochman) [inline config/routes.rb]: same here
why do you need the new action?

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    it 'return status 200' do
+      expect(response.status).to eq 200
+    end
+
+    it 'returns success message' do
+      expect(response_json['message']).to eq 'Your payment was successful'
+    end
+
+    it 'user role changes to subscriber when payment is successful' do 
+      allow(new_subscriber).to receive(:role).and_return('subscriber')
```
Comment (oliverochman) [inline spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: can you explain this line?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/subscriptions_controller.rb (hunk 1 above)
Comment (yasmineezequiel) [reply app/controllers/api/v1/subscriptions_controller.rb]: Deleted

File: config/routes.rb (hunk 2 above)
Comment (yasmineezequiel) [reply config/routes.rb]: deleted

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb (hunk 3 above)
Comment (yasmineezequiel) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: We took from the ATM challenge this line of code so we could update the new_subscriber Model

File: Gemfile [hunk 4]
```diff
@@ -11,6 +11,8 @@ gem 'rack-cors'
 gem 'active_model_serializers'
 gem 'devise_token_auth'
 gem 'pundit', '~> 1.1'
+gem 'stripe-rails'
+gem 'stripe-ruby-mock', '~> 2.5', require: 'stripe_mock'
```
Comment (tochman) [inline Gemfile]: Please move this out of the production section. This gem is only used in development and test

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb (hunk 3 above)
Comment (tochman) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: It's not really relevant here

File: spec/requests/api/v1/sessions_spec.rb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 RSpec.describe 'Sessions', type: :request do
-  let(:user) { FactoryBot.create(:user) }
+  let(:user) { FactoryBot.create(:user, role: 'subscriber') }
```
Comment (tochman) [inline spec/requests/api/v1/sessions_spec.rb]: You can omit the module name here (`FactoryBot`)

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb (hunk 3 above)
Comment (oliverochman) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: you are not really testing the request and its response here. Feels more like a unit spec

Comment (piavW) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: But we want to make sure that when a payment is successful the user roles is changed to subscriber. That's part of the feature and thus within scope.

File: Gemfile (hunk 4 above)
Comment (piavW) [reply Gemfile]: So stripe ruby mock in development test and stripe-rails in development then?

Comment (piavW) [reply Gemfile]: Done.

**Review by DIzosimov — APPROVED** [review]

**Review by piavW — APPROVED** [review]

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb (hunk 3 above)
Comment (tochman) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: Well, then pull out the user from the db and assert that he is a subscriber.

File: spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      other_subscriber.update_attribute(:role, 'subscriber') 
+      expect(other_subscriber.subscriber?).to eq true
+    end
+  end
+
+  describe 'Payment is declined' do
+    describe 'because payment lacks stripe token' do
+      before do
+        post '/api/v1/subscriptions',
+        params: { strikeToken: nil },
```
Comment (tochman) [inline spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: It should probably be `stripeToken`, right? Not `strike`

Comment (piavW) [reply spec/requests/api/v1/visitor_can_pay_for_subscription_spec.rb]: Fixed

**Review by faraznaeem — APPROVED** [review]

**Comment by faraznaeem** [comment]
Please go over the story and check off the AC

**Review by oliverochman — APPROVED** [review]

**Pull request merged**